### PR TITLE
Knowledge

### PR DESCRIPTION
**Pull request by Tigul**
# Description 
Adds the new Knowledge and Reasoing engine

### CODE REVIEW

File: doc/source/knowledge_examples.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+=======================
+Knowledge Examples
+=======================
```
Comment (tomsch420) [inline doc/source/knowledge_examples.rst:3]: is the inclusion of the example missing here?

File: src/pycram/datastructures/partial_designator.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    A partial designator_description is somewhat between a DesignatorDescription and a specified designator_description. Basically it is a
+    partially initialized specified designator_description which can take a list of input arguments (like a DesignatorDescription)
+    and generate a list of specified designators with all possible permutations of the input arguments.
+
+    PartialDesignators are designed as generators, as such they need to be iterated over to yield the possible specified
+    designators. Please also keep in mind that at the time of iteration all parameter of the specified designator_description need
+    to be filled, otherwise a TypeError will be raised, see the example below for usage.
+
+    .. code-block:: python
+            # Example usage
```
Comment (tomsch420) [inline src/pycram/datastructures/partial_designator.py:25]: shouldnt example usage be prefixed with
>>> my_boject = ...
?

File: src/pycram/datastructures/partial_designator.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        for args_combination in PartialDesignator.generate_permutations(self.args):
+            for kwargs_combination in PartialDesignator.generate_permutations(self.kwargs.values()):
+                yield self.performable(*args_combination, **dict(zip(self.kwargs.keys(), kwargs_combination)))
+
+    @staticmethod
+    def generate_permutations(args: Iterable) -> List:
+        """
+        Generates all possible permutations of the given arguments. This uses itertools.product to generate the
+        permutations.
```
Comment (tomsch420) [inline src/pycram/datastructures/partial_designator.py:75]: unnescessary information in the docstring

Comment (tomsch420) [reply src/pycram/datastructures/partial_designator.py:75]: write cartesian product or something that indicates what you are doing conceptually. If i want to know how you do it in detail i can read the code

File: src/pycram/datastructures/property.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        """
+        success = self.result
+        for node in PreOrderIter(self.root):
+            success &= node.result.success
+        return success
+
+    @property
+    def resolved(self) -> bool:
+        """
+        Returns True if all nodes in the tree are resolved. This is done by iterating over the tree and checking if all
```
Comment (tomsch420) [inline src/pycram/datastructures/property.py:103]: TMI

File: src/pycram/datastructures/property.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+        :param properties: A list of properties to which are the children of this node
+        """
+        super().__init__(parent=None, children=properties)
+        self.result = True
+
+    def simplify(self):
+        """
+        Simplifies the tree structure by merging nodes of the same type. This is done by iterating over the tree and
+        merging nodes of the same type.
```
Comment (tomsch420) [inline src/pycram/datastructures/property.py:134]: TMI and duplicated?

File: src/pycram/datastructures/property.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        :return: Returns the root node of the tree
+        """
+        for node in PreOrderIter(self.root):
+            for child in node.children:
+                if type(node) is type(child):
+                    self.merge_nodes(node, child)
+        return self.root
+
+    @staticmethod
+    def merge_nodes(node1: Node, node2: Node) -> None:
```
Comment (tomsch420) [inline src/pycram/datastructures/property.py:145]: i find the word "merging" for stealing the children quite confusing

File: src/pycram/datastructures/property.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        :param node2: Node which children should be appended to node1 and then deleted
+        """
+        node2.parent = None
+        node1.children = node2.children + node1.children
+
+
+class And(PropertyOperator):
+    """
+    Class to represent a logical and operator in a tree structure. This class inherits from PropertyOperator and adds a
+    method to evaluate the children as an and operator.
```
Comment (tomsch420) [inline src/pycram/datastructures/property.py:160]: I am very unsure about the information thats given here. Like Inherits from is written in the class definition and the adds a method stuff down there too. Also it is required by the super class.

File: src/pycram/designators/action_designator.py
Comment (tomsch420) [inline src/pycram/designators/action_designator.py:1]: Git shows this in a weird version. This file just got restructured didnt it?

File: src/pycram/knowledge/__init__.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# from ..datastructures.knowledge_engine import KnowledgeEngine
+#
+# import os
+# import sys
+
+# path = os.path.dirname(os.path.abspath(__file__))
```
Comment (tomsch420) [inline src/pycram/knowledge/__init__.py:6]: delete if this is unwanted

File: src/pycram/knowledge/knowledge_engine.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+from typeguard import check_type, TypeCheckError
+
+from ..datastructures.partial_designator import PartialDesignator
+from ..datastructures.property import Property, ResolvedProperty, PropertyOperator
+from .knowledge_source import KnowledgeSource
+from typing_extensions import Type, Callable, List, TYPE_CHECKING, Dict, Any
+
+from ..failures import KnowledgeNotAvailable, ReasoningError
+# This import is needed since the subclasses of KnowledgeSource need to be imported to be known at runtime
+from .knowledge_sources import *
```
Comment (tomsch420) [inline src/pycram/knowledge/knowledge_engine.py:16]: if stuff like this is needed i recommend to import it in the __init__.py I got the same problems for the subclass json serializer but the best way to resolve the conflict for me was  to import what i can in the init of a module and the rest (cyclic dependency stuff) has to be done by the user when he wants to use that class.

File: src/pycram/knowledge/knowledge_sources/__init__.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+from os.path import dirname, basename, isfile, join
+import glob
+
+modules = glob.glob(join(dirname(__file__), "*.py"))
+__all__ = [basename(f)[:-3] for f in modules if isfile(f) and not f.endswith('__init__.py')]
```
Comment (tomsch420) [inline src/pycram/knowledge/knowledge_sources/__init__.py:5]: this looks fishy, please explain

**Review by tomsch420 — CHANGES_REQUESTED** [review]
Generally good concept and nice clean up of actions.

**Review by tomsch420 — APPROVED** [review]

File: doc/source/knowledge_examples.rst (hunk 1 above)
Comment (Tigul) [reply doc/source/knowledge_examples.rst:3]: No, to have categories in the toc you need a page under which you can put other pages that's what this is for,.

File: src/pycram/datastructures/property.py (hunk 6 above)
Comment (Tigul) [reply src/pycram/datastructures/property.py:155]: But that's exacty what it does, it combines two nodes by merging them

File: src/pycram/knowledge/knowledge_sources/__init__.py (hunk 10 above)
Comment (Tigul) [reply src/pycram/knowledge/knowledge_sources/__init__.py:5]: This imports all files in this directory to ensure that all knowledge sources are known.

**Pull request merged**